dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/DynamicPureLoopCutting.java

```diff
@@ -130,6 +133,11 @@ private void insertSideEffectChecks(IterationData iter) {
             trackingRegs.add(trackingReg);
 
             final Event execCheck = EventFactory.newExecutionStatus(trackingReg, sideEffect);
+            // DynamicPureLoopCutting is run after unrolling, thus we need to assign all
+            // three of oId, cId and uId.
+            execCheck.setOId(insertionPoint.getOId());
+            execCheck.setCId(insertionPoint.getCId());
+            execCheck.setUId(insertionPoint.getUId());
```

Comment:
I'm not a big fan of this. Unlike the other cases where you copied metadata over, in this case the inserted event does not have any reasonable association with the metadata you attach to it.
This totally breaks the meaning of what the different Ids even mean!

I know that you want to have this to do "coverage" so that each branch representative has reasonable metadata that connects it to the parsed code.
But rather than destroying the meaning of our metadata to fix this problem for coverage reports, you should rather fix the coverage reporting itself. For example, if the canonical representative of a branch has no metadata (cause we created it via some passes), then maybe just use a different event in the branch to distinguish it from other branches (i.e. compute a mapping `Branch->Event` that gives you "good" representatives).

Reply:
I agree that the association between `execCheck` and `insertionPoint` is less clear than in other cases (to the point that I initially had a hard time to decide from which event to copy the metadata). 

That being said, I disagree that this destroys the meaning of the metadata. I don't think we have any case where we created an event `a` that is not related in any form with any other event `b` (as soon as we have `b`, we can copy its metadata). For me is actually the opposite: allowing several events to have some id with value `-1` after all processing destroys the meaning of the metadata. My interpretation of ids is that they are unique identifiers and the meaning of `-1` is "it has not yet been assigned". 

Based on all the above, I would consider that `execCheck` is related to `spinloopStart` and copy metadata from it.

Reply:
Regarding the definition of the `ids`: they are defined as the unique global id an event had at some previous point in time, i.e, they track the history of an event over time. The nonexistance of an id then means that the event simply did not exist in any shape or form. In this sense, the value "-1" actually has meaning. because it indicates nonexistence.
Now how do you distinguish multiple events with the same "-1" entries? Well, you just don't compare them on these id's but rather on `globalId` (this the only unique id!). 
If you compare events on, e.g., their `oIds`, then if exactly one of the two events has an `-1` entry, then the events have no common history. If both events have `-1` entries then you, need to check their other ids to see if they have anything in common.

Regarding the "metadata" aspect: Metadata is "additional/optional information", i.e., under ideal circumstance it should be droppable without affecting semantic/functional correctness (our code certainly has places that strictly require metadata, but I think that is a fault in the code). In the case of coverage reports for example, one possible way to achieve this is to put events with `oId=-1` into separate groups(*), i.e., singleton groups. 
Then even if we dropped all our metadata, the report would still be working, but just more fine-grained as now everything is distinguished from one another.
While I'm on this topic: Why don't we just filter out all `oId=-1` and `cLine=-1` events? The former indicates that the event did not exist during parsing, so it surely has no syntactic counterpart in the code that it represents.

>Based on all the above, I would consider that `execCheck` is related to `spinloopStart` and copy metadata from it.

`spinloopStart` is a reasonable choice. Intuitively, I would say that the most appropriate choice (if any), would be to match the metadata of the extra events (both the `assumeSideEffect` and the `execCheck`) to the backjump of the loop.
Unfortunately, I'm not too sure if we retain the backjump metadata anywhere except possibly the very last iteration where we insert the bound. The second best choice is the loop header label (`spinloopStart`) and I can live with that :).


(*) If you want to be more precise in such cases, you could group events by their other ids: if `oId` fails, try `cId`. If that fails, try `uId`, and if that fails try `globalId` (which will always be able to distinguish events by its uniqueness).